lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Shared.hs

```diff
@@ -0,0 +1,311 @@
+      , pendingSharedStateAddressPoolGap :: !AddressPoolGap
+          -- ^ Address pool gap to be used in the address pool of shared state
+      }
+    | SharedState
+      { sharedStateDerivationPrefix :: !DerivationPrefix
+        -- ^ Derivation path prefix from a root key up to the account key
+      , sharedStateAddressPool :: !(AddressPool 'MultisigScript k)
+        -- ^ Address pool tracking the shared addresses. Co-owning is based on
+        -- payment credential only. Moreover, the parent context information is stored
+        -- ie., validated script template for payment credential, optional
+        -- validated script template for delegation credential and account public key with which
+        -- the shared wallet was initiated
+      } deriving (Generic)
+
+deriving instance
+    ( Show (k 'AccountK XPub)
+    ) => Show (SharedState n k)
+
+deriving instance
+    ( Eq (k 'AccountK XPub)
+    ) => Eq (SharedState n k)
+
+instance
+    ( NFData (k 'AccountK XPub)
+    ) => NFData (SharedState n k)
+
+-- | Purpose for shared wallets is a constant set to 1854' (or 0x8000073E) following the original
+-- CIP-1854 Multi-signature Wallets.
+--
+-- It indicates that the subtree of this node is used according to this
+-- specification.
+--
+-- Hardened derivation is used at this level.
+purposeCIP1854 :: Index 'Hardened 'PurposeK
+purposeCIP1854 = toEnum 0x8000073E
+
+-- | Create a pending SharedState from the all needed ingredients.
+-- There is no validation and it is unsafe way.
+unsafePendingSharedState
+    :: forall (n :: NetworkDiscriminant) k. k 'AccountK XPub
+    -> Index 'Hardened 'AccountK
+    -> AddressPoolGap
+    -> ScriptTemplate
+    -> Maybe ScriptTemplate
+    -> SharedState n k
+unsafePendingSharedState accXPub accIx g pTemplate dTemplateM =
+    PendingSharedState
+    { pendingSharedStateDerivationPrefix = DerivationPrefix ( purposeCIP1854, coinTypeAda, accIx )
+    , pendingSharedStateAccountKey = accXPub
+    , pendingSharedStatePaymentTemplate = pTemplate
+    , pendingSharedStateDelegationTemplate = dTemplateM
+    , pendingSharedStateAddressPoolGap = g
+    }
+
+-- | Create a new SharedState.
+newSharedState
+    :: forall (n :: NetworkDiscriminant) k.
+    ( MkKeyFingerprint k Address
+    , SoftDerivation k
+    , Typeable n
+    , MkKeyFingerprint k (Proxy n, k 'AddressK XPub) )
+    => k 'AccountK XPub
+    -> Index 'Hardened 'AccountK
+    -> AddressPoolGap
+    -> ScriptTemplate
+    -> Maybe ScriptTemplate
+    -> SharedState n k
+newSharedState accXPub accIx g pTemplate dTemplateM =
+    let pendingSharedState = unsafePendingSharedState accXPub accIx g pTemplate dTemplateM
+    in trySharedState pendingSharedState
+
+trySharedState
+    :: forall (n :: NetworkDiscriminant) k.
+     ( MkKeyFingerprint k (Proxy n, k 'AddressK XPub)
+     , MkKeyFingerprint k Address
+     , Typeable n
+     , SoftDerivation k )
+    => SharedState n k
+    -> SharedState n k
+trySharedState = \case
```

Comment:
I am not sure what this function does (turn a "pending" into an active state or identity if already active?)? Maybe the intent could be clarified with a better name?

Reply:
```turn a "pending" into an active state or identity if already active?```` yes